File: typegate/tests/runtimes/graphql/typegraphs/deno/graphql.ts
// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
// SPDX-License-Identifier: Elastic-2.0

import { Policy, t, typegraph } from "@typegraph/sdk/index.js";
import { GraphQLRuntime } from "@typegraph/sdk/runtimes/graphql.js";
import * as effects from "@typegraph/sdk/effects.js";
import { PrismaRuntime } from "@typegraph/sdk/providers/prisma.js";

const user = t.struct(
  {
    id: t.string(),
    name: t.string(),
    // TODO more fields with more types
  },
  { name: "User" },
);

const createUserInput = t.struct({
  name: t.string(),
  username: t.string(),
  email: t.string(),
});

export const tg = await typegraph("graphql", (g) => {
  const graphql = new GraphQLRuntime("https://graphqlzero.almansi.me/api");
  const pub = Policy.public();
  const db = new PrismaRuntime("graphql", "POSTGRES");

  const message = t.struct(
    {
      id: t.integer({}, { asId: true, config: { auto: true } }),
      title: t.string(),
      user_id: t.string({}, { name: "uid" }),
      user: graphql.query(
        t.struct({
          id: t.string({}, { asId: true }).fromParent("uid"),
        }),
        t.optional(user),
      ),
    },
    { name: "message" },
  );

  g.expose({
    user: graphql
      .query(t.struct({ id: t.string({}, { asId: true }) }), user)
      .withPolicy(pub),
    users: graphql
      .query(
        t.struct({}),
        t.struct({
          data: t.list(user),
        }),
      )
      .withPolicy(pub),
    create_user: graphql
      .mutation(
        t.struct({
          input: createUserInput,
        }),
        user,
        effects.create(false),
      )
      .withPolicy(pub),
    create_message: db.create(message).withPolicy(pub),
    messages: db.findMany(message).withPolicy(pub),
  });
});

File: tests/runtimes/graphql/typegraphs/python/graphql.py
from typegraph.providers.prisma import PrismaRuntime
from typegraph.runtimes.graphql import GraphQLRuntime

from typegraph import Graph, Policy, effects, t, typegraph

_user = t.struct(
    {"id": t.string(as_id=True), "name": t.string()},
    name="User",
)

create_user_input = t.struct(
    {
        "name": t.string(),
        "username": t.string(),
        "email": t.string(),
    },
    name="CreateUserInput",
)


@typegraph()
def graphql(g: Graph):
    graphql = GraphQLRuntime("https://graphqlzero.almansi.me/api")
    public = Policy.public()
    db = PrismaRuntime("graphql", "POSTGRES")

    message = t.struct(
        {
            "id": t.integer(as_id=True, config=["auto"]),
            "title": t.string(),
            "user_id": t.string(name="uid"),
            "user": graphql.query(
                t.struct({"id": t.string(as_id=True).from_parent("uid")}),
                t.optional(_user),
            ),
        },
        name="message",
    )

    g.expose(
        user=graphql.query(
            t.struct({"id": t.string(as_id=True)}),
            _user,
        ).with_policy(public),
        users=graphql.query(
            t.struct({}),
            t.struct(
                {"data": t.list(_user)},
            ),
        ).with_policy(public),
        createUser=graphql.mutation(
            t.struct(
                {
                    "input": create_user_input,
                }
            ),
            _user,
            effects.create(False),
        ).with_policy(public),
        create_message=db.create(message).with_policy(public),
        messages=db.find_many(message).with_policy(public),
    )
